src/passes/MemoryPacking.cpp

```diff
@@ -23,151 +38,607 @@
-        if (segment.isPassive) {
-          return;
+    Index firstNewIndex = packed.size();
+    size_t segmentsRemaining = segments.size() - origIndex;
+    createSplitSegments(builder, segment, ranges, packed, segmentsRemaining);
+    createReplacements(
+      module, ranges, currReferrers, replacements, firstNewIndex);
+  }
+
+  segments.swap(packed);
+
+  if (module->features.hasBulkMemory()) {
+    replaceBulkMemoryOps(runner, module, replacements);
+  }
+}
+
+bool MemoryPacking::canSplit(const Memory::Segment& segment,
+                             const std::vector<Expression*>& referrers) {
+  if (segment.isPassive) {
+    for (auto* referrer : referrers) {
+      if (auto* init = referrer->dynCast<MemoryInit>()) {
+        // Do not try to split if there is a nonconstant offset or size
+        if (!init->offset->is<Const>() || !init->size->is<Const>()) {
+          return false;
         }
       }
     }
+    return true;
+  } else {
+    // Active segments can only be split if they have constant offsets
+    return segment.offset->is<Const>();
+  }
+}
 
-    std::vector<Memory::Segment> packed;
+void MemoryPacking::calculateRanges(const Memory::Segment& segment,
+                                    const std::vector<Expression*>& referrers,
+                                    std::vector<Range>& ranges) {
+  auto& data = segment.data;
+  if (data.size() == 0) {
+    return;
+  }
 
-    // we can only handle a constant offset for splitting
-    auto isSplittable = [&](const Memory::Segment& segment) {
-      return segment.offset->is<Const>();
-    };
+  // Calculate initial zero and nonzero ranges
+  size_t start = 0;
+  while (start < data.size()) {
+    size_t end = start;
+    while (end < data.size() && data[end] == 0) {
+      end++;
+    }
+    if (end > start) {
+      ranges.push_back({true, start, end});
+      start = end;
+    }
+    while (end < data.size() && data[end] != 0) {
+      end++;
+    }
+    if (end > start) {
+      ranges.push_back({false, start, end});
+      start = end;
+    }
+  }
 
-    for (auto& segment : module->memory.segments) {
-      if (!isSplittable(segment)) {
-        packed.push_back(segment);
+  // Number of adjacent interior zeroes for which splitting is beneficial
+  size_t threshold = 0;
+  if (segment.isPassive) {
+    // Passive segment metadata size
+    threshold += 2;
+    // Zeroes on the edge do not increase the number of segments or data.drops,
+    // so their threshold is lower. The threshold for interior zeroes depends on
+    // an estimate of the number of new memory.fill and data.drop instructions
+    // splitting would introduce.
+    size_t edgeThreshold = 0;
+    for (auto* referrer : referrers) {
+      if (referrer->is<MemoryInit>()) {
+        // Splitting adds a new memory.fill and a new memory.init
+        threshold += MEMORY_FILL_SIZE * 2;
+        edgeThreshold += MEMORY_FILL_SIZE;
+      } else {
+        threshold += DATA_DROP_SIZE;
```

Comment:
Sorry I still have hard time understanding why thresholds are computed this way. Why would this depend on the number of original referrers? Why `threshold` is increased by `MEMORY_FILL_SIZE * 2` for `MemoryInit`? The number of split isn't something separate from the minimum length of a segment?

Reply:
Talked in person and now I understand this part better, thanks :) I'll summarize what we talked about. The points I was confused about were twofold:
- Why is that not `threshold += MEMORY_INIT_SIZE + MEMORY_FILL_SIZE` but `MEMORY_FILL_SIZE * 2`? Turned out two of them are the same size. Maybe defining `MEMORY_INIT_SIZE` with the same amount would help readability.
- I couldn't understand why every existing `memory.init` would contribute to the size increase every time we cut a new segment, because all `memory.init` refer different parts of the segment. This calculation is an approximation, and the reason it's a good approximation is usually in the toolchain we generate a single big `memory.init` that covers a whole segment. Clarifying that this is an approximation in the comment might help.

Reply:
Fixed, and as @kripken pointed out above, they actually are not the same size. Oops!